velox/functions/prestosql/json/SimdJsonExtractor.cpp

```diff
@@ -0,0 +1,590 @@
+
+#include <cctype>
+#include <unordered_map>
+#include <vector>
+
+#include "boost/algorithm/string/trim.hpp"
+#include "folly/String.h"
+#include "folly/json.h"
+#include "simdjson.h"
+#include "velox/common/base/Exceptions.h"
+#include "velox/functions/prestosql/json/JsonPathTokenizer.h"
+
+namespace facebook::velox::functions {
+
+namespace {
+
+class SimdJsonExtractor {
+ public:
+  // Use this method to get an instance of SimdJsonExtractor given a json path.
+  static SimdJsonExtractor& getInstance(folly::StringPiece path) {
+    // Pre-process
+    auto trimedPath = folly::trimWhitespace(path).str();
+
+    std::shared_ptr<SimdJsonExtractor> op;
+    if (kExtractorCache.count(trimedPath)) {
+      op = kExtractorCache.at(trimedPath);
+    } else {
+      if (kExtractorCache.size() == kMaxCacheNum) {
+        // TODO: Blindly evict the first one, use better policy
+        kExtractorCache.erase(kExtractorCache.begin());
+      }
+      op = std::make_shared<SimdJsonExtractor>(trimedPath);
+      kExtractorCache[trimedPath] = op;
+    }
+    return *op;
+  }
+
+  std::optional<std::string> extract(const std::string& json);
+  std::optional<std::string> extractScalar(const std::string& json);
+  std::optional<std::string> extractFromObject(
+      int path_index,
+      simdjson::dom::object obj);
+  std::optional<std::string> extractFromArray(
+      int path_index,
+      simdjson::dom::array arr);
+  std::optional<std::string> extractOndemand(const std::string& json);
+  std::optional<std::string> extractKeysOndemand(const std::string& json);
+  std::optional<std::string> extractFromObjectOndemand(
+      int path_index,
+      simdjson::ondemand::object obj);
+  std::optional<std::string> extractFromArrayOndemand(
+      int path_index,
+      simdjson::ondemand::array arr);
+  bool isDocBasicType(simdjson::ondemand::document& doc);
+  bool isValueBasicType(
+      simdjson::simdjson_result<simdjson::ondemand::value> rlt);
+
+  // Shouldn't instantiate directly - use getInstance().
+  explicit SimdJsonExtractor(const std::string& path) {
+    if (!tokenize(path)) {
+      VELOX_USER_FAIL("Invalid JSON path: {}", path);
+    }
+  }
+
+ private:
+  bool tokenize(const std::string& path) {
+    if (path.empty()) {
+      return false;
+    }
+    if (!kTokenizer.reset(path)) {
+      return false;
+    }
+
+    while (kTokenizer.hasNext()) {
+      if (auto token = kTokenizer.getNext()) {
+        tokens_.push_back(token.value());
+      } else {
+        tokens_.clear();
+        return false;
+      }
+    }
+    return true;
+  }
+
+  // Cache tokenize operations in SimdJsonExtractor across invocations in the
+  // same thread for the same JsonPath.
+  thread_local static std::
+      unordered_map<std::string, std::shared_ptr<SimdJsonExtractor>>
+          kExtractorCache;
+  thread_local static JsonPathTokenizer kTokenizer;
+
+  // Max extractor number in extractor cache
+  static const uint32_t kMaxCacheNum{32};
```

Comment:
Would you update PR description to describe the logic around this cache?

Reply:
reference to the implementation of folly extract temporarily.

Reply:
Thank you for explaining. Instead of doing that we could add initialize method to the function taking jsonPath. This method is called once per thread of execution, it can parse the path and save it in a member variable. The 'call' method can then access already parsed path.

See https://facebookincubator.github.io/velox/develop/scalar-functions.html#access-to-session-properties-and-constant-inputs

Reply:
Thanks for the guidance, I'm confused about whether different functions can use the member variable(already parsed path). In the current implementation, already parsed paths are shared by different functions, like simdJsonSize/simdJsonExtractScalar.

Reply:
done